src/objectid.ts

```diff
@@ -104,27 +125,34 @@ export class ObjectId extends BSONValue {
     }
 
     // The following cases use workingId to construct an ObjectId
-    if (workingId == null || typeof workingId === 'number') {
+    if (typeof workingId === 'string') {
+      if (option === OID_SKIP_VALIDATE) {
+        this.__id = workingId;
+      } else {
+        const validString = ObjectId.validateHexString(workingId);
+        if (validString) {
+          this.__id = validString;
+        } else {
+          throw new BSONError(
+            'input must be a 24 character hex string, 12 byte Uint8Array, or an integer'
+          );
+        }
+      }
+    } else if (workingId == null || typeof workingId === 'number') {
       // The most common use case (blank id, new objectId instance)
       // Generate a new id
-      this.buffer = ObjectId.generate(typeof workingId === 'number' ? workingId : undefined);
-    } else if (ArrayBuffer.isView(workingId) && workingId.byteLength === 12) {
+      this.__id = ObjectId.generate(typeof workingId === 'number' ? workingId : undefined);
+    } else if (ArrayBuffer.isView(workingId)) {
```

Comment:
This change goes along with the change to deserialization code [here](https://github.com/mongodb/js-bson/pull/703/files#diff-8319784c1d6bd6d2027c93426c54a9dafdf11e056b9178f1d6bd06d675ff8c77R266)

Instead of allocating a new UInt8Array(12), copying over the 12 bytes from the parent buffer, then passing in that new buffer, only to convert it to a string and throw away that temporary buffer. You can pass in a buffer and an offset to `new ObjectId(buffer, offset)` and it will just grab the next 12 bytes after that offset and encode that to hex. 

It's probably worth enforcing the buffer size === 12 **if** no offset was passed in, also if `offset + 12 > byteLength`. WDYT?

Reply:
> It's probably worth enforcing the buffer size === 12 if no offset was passed in, also if offset + 12 > byteLength. WDYT?

Yea those sound like the correct assertions. It's important that when inspecting the ObjectId you get the same visual results as you do when serializing. So whether it is a view on a larger arrayBuffer or exactly 12 bytes, we just do not want to permit an OID to carry around a buffer that implies more or less than exactly that amount of data.